cmd/policy.go

```diff
@@ -70,42 +89,61 @@ func (sys *PolicySys) IsAllowed(args policy.Args) bool {
 	return args.IsOwner
 }
 
-// Init - initializes policy system from policy.json of all buckets.
-func (sys *PolicySys) Init(objAPI ObjectLayer) error {
-	if objAPI == nil {
-		return errInvalidArgument
-	}
-
+// Refresh in-memory bucket policy cache.
+func (sys *PolicySys) refreshInmemoryBucketpolicy(objAPI ObjectLayer) {
 	buckets, err := objAPI.ListBuckets(context.Background())
 	if err != nil {
-		return err
+		return
 	}
-
+	sys.RemoveDeletedBuckets(buckets)
 	for _, bucket := range buckets {
 		config, err := GetPolicyConfig(objAPI, bucket.Name)
 		if err != nil {
-			if _, ok := err.(BucketPolicyNotFound); !ok {
-				return err
+			if _, ok := err.(ObjectNotFound); ok {
+				// To delete in memory cache of the policy.
+				sys.Set(bucket.Name, policy.Policy{Version: policy.DefaultVersion})
 			}
-		} else {
-			// This part is specifically written to handle migration
-			// when the Version string is empty, this was allowed
-			// in all previous minio releases but we need to migrate
-			// those policies by properly setting the Version string
-			// from now on.
-			if config.Version == "" {
-				logger.Info("Found in-consistent bucket policies, Migrating them for Bucket: (%s)", bucket.Name)
-				config.Version = policy.DefaultVersion
-
-				if err = savePolicyConfig(objAPI, bucket.Name, config); err != nil {
-					return err
-				}
+			continue
+		}
+		// This part is specifically written to handle migration
+		// when the Version string is empty, this was allowed
+		// in all previous minio releases but we need to migrate
+		// those policies by properly setting the Version string
+		// from now on.
+		if config.Version == "" {
+			logger.Info("Found in-consistent bucket policies, Migrating them for Bucket: (%s)", bucket.Name)
+			config.Version = policy.DefaultVersion
+
+			if err = savePolicyConfig(objAPI, bucket.Name, config); err != nil {
+				continue
 			}
-
-			sys.Set(bucket.Name, *config)
 		}
+		sys.Set(bucket.Name, *config)
+	}
+}
+
+// Init - initializes policy system from policy.json of all buckets.
+func (sys *PolicySys) Init(objAPI ObjectLayer) error {
+	if objAPI == nil {
+		return errInvalidArgument
 	}
 
+	// Refresh in-memory cache once during boot.
+	sys.refreshInmemoryBucketpolicy(objAPI)
+
+	ticker := time.NewTicker(globalRefreshBucketPolicyInterval)
+	// Refresh in-memory cache in background.
```

Comment:
My choice is not in-memory because of `PolicySys` itself is in-memory.  initialize PolicySys is more appropriate IMO